File: products/LoanHub/apps/backend/database/schemas/finance.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from database.models.enums import PaymentDirection, PaymentMethod, PaymentProvider


class BorrowerRequestFeeWrite(BaseModel):
    company_id: UUID | None = None
    name: str = Field(default="Borrow request service fee", min_length=3, max_length=160)
    fee_type: str = Field(default="flat", pattern="^(flat|percentage|hybrid)$")
    flat_amount: Decimal = Field(default=Decimal("0"), ge=0, max_digits=15, decimal_places=2)
    percentage: Decimal = Field(default=Decimal("0"), ge=0, le=100, max_digits=10, decimal_places=6)
    minimum_amount: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    maximum_amount: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    currency: str = Field(default="LSL", pattern="^[A-Z]{3}$")
    required_before_submission: bool = True
    refundable: bool = False
    effective_from: datetime | None = None
    effective_to: datetime | None = None
    is_active: bool = True

    @model_validator(mode="after")
    def validate_amounts(self):
        if self.is_active and self.fee_type == "flat" and self.flat_amount <= 0:
            raise ValueError("An active flat fee must be greater than zero")
        if self.is_active and self.fee_type == "percentage" and self.percentage <= 0:
            raise ValueError("An active percentage fee must be greater than zero")
        if self.is_active and self.fee_type == "hybrid" and self.flat_amount <= 0 and self.percentage <= 0:
            raise ValueError("An active hybrid fee needs a flat amount or percentage")
        if self.minimum_amount is not None and self.maximum_amount is not None and self.minimum_amount > self.maximum_amount:
            raise ValueError("Minimum amount cannot exceed maximum amount")
        return self


class BorrowerRequestFeeRead(BorrowerRequestFeeWrite):
    id: UUID
    created_by_user_id: UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AccountOpeningFeeWrite(BaseModel):
    company_id: UUID | None = None
    name: str = Field(default="Assisted borrower account opening fee", min_length=3, max_length=160)
    fee_type: str = Field(default="flat", pattern="^(flat|percentage|hybrid)$")
    flat_amount: Decimal = Field(default=Decimal("0"), ge=0, max_digits=15, decimal_places=2)
    percentage: Decimal = Field(default=Decimal("0"), ge=0, le=100, max_digits=10, decimal_places=6)
    minimum_amount: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    maximum_amount: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    currency: str = Field(default="LSL", pattern="^[A-Z]{3}$")
    effective_from: datetime | None = None
    effective_to: datetime | None = None
    is_active: bool = True

    @model_validator(mode="after")
    def validate_amounts(self):
        if self.is_active and self.fee_type == "flat" and self.flat_amount <= 0:
            raise ValueError("An active flat fee must be greater than zero")
        if self.is_active and self.fee_type == "percentage" and self.percentage <= 0:
            raise ValueError("An active percentage fee must be greater than zero")
        if self.is_active and self.fee_type == "hybrid" and self.flat_amount <= 0 and self.percentage <= 0:
            raise ValueError("An active hybrid fee needs a flat amount or percentage")
        if self.minimum_amount is not None and self.maximum_amount is not None and self.minimum_amount > self.maximum_amount:
            raise ValueError("Minimum amount cannot exceed maximum amount")
        return self


class AccountOpeningFeeRead(AccountOpeningFeeWrite):
    id: UUID
    created_by_user_id: UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TransactionAgreementWrite(BaseModel):
    company_id: UUID
    name: str = Field(default="Platform cash transaction charge agreement", min_length=3, max_length=180)
    inbound_percentage: Decimal = Field(default=Decimal("0"), ge=0, le=100, max_digits=10, decimal_places=6)
    outbound_percentage: Decimal = Field(default=Decimal("0"), ge=0, le=100, max_digits=10, decimal_places=6)
    inbound_flat_fee: Decimal = Field(default=Decimal("0"), ge=0, max_digits=15, decimal_places=2)
    outbound_flat_fee: Decimal = Field(default=Decimal("0"), ge=0, max_digits=15, decimal_places=2)
    minimum_charge: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    maximum_charge: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    currency: str = Field(default="LSL", pattern="^[A-Z]{3}$")
    settlement_frequency: str = Field(default="monthly", pattern="^(daily|weekly|monthly|quarterly|custom)$")
    settlement_day: int | None = Field(default=None, ge=1, le=31)
    effective_from: date
    effective_to: date | None = None
    terms: str | None = Field(default=None, max_length=20000)

    @model_validator(mode="after")
    def validate_period(self):
        if self.effective_to is not None and self.effective_to < self.effective_from:
            raise ValueError("Effective-to date cannot be before effective-from")
        if self.minimum_charge is not None and self.maximum_charge is not None and self.minimum_charge > self.maximum_charge:
            raise ValueError("Minimum charge cannot exceed maximum charge")
        return self


class TransactionAgreementRead(TransactionAgreementWrite):
    id: UUID
    agreement_number: str
    status: str
    owner_accepted_by_user_id: UUID | None
    company_accepted_by_user_id: UUID | None
    owner_accepted_at: datetime | None
    company_accepted_at: datetime | None
    activated_at: datetime | None
    suspended_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AgreementDecision(BaseModel):
    accept: bool = True
    reason: str | None = Field(default=None, max_length=1000)


class ChargeLedgerRead(BaseModel):
    id: UUID
    company_id: UUID
    agreement_id: UUID
    payment_id: UUID
    claim_id: UUID | None
    direction: PaymentDirection
    provider: PaymentProvider
    payment_purpose: str
    gross_amount: Decimal
    percentage_rate: Decimal
    flat_fee: Decimal
    charge_amount: Decimal
    currency: str
    status: str
    accrued_at: datetime
    claimed_at: datetime | None
    settled_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ChargeClaimCreate(BaseModel):
    company_id: UUID
    period_start: date
    period_end: date
    due_days: int = Field(default=14, ge=0, le=365)
    notes: str | None = Field(default=None, max_length=5000)

    @model_validator(mode="after")
    def validate_period(self):
        if self.period_end < self.period_start:
            raise ValueError("Claim period end cannot be before its start")
        return self


class ChargeClaimRead(BaseModel):
    id: UUID
    company_id: UUID
    agreement_id: UUID
    claim_number: str
    period_start: date
    period_end: date
    transaction_count: int
    gross_transaction_value: Decimal
    amount: Decimal
    currency: str
    status: str
    issued_at: datetime | None
    due_at: datetime | None
    acknowledged_at: datetime | None
    paid_at: datetime | None
    disputed_at: datetime | None
    notes: str | None
    dispute_reason: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ClaimCompanyDecision(BaseModel):
    action: str = Field(pattern="^(acknowledge|dispute)$")
    reason: str | None = Field(default=None, max_length=2000)


class ClaimCashSettlementCreate(BaseModel):
    payment_method: PaymentMethod = PaymentMethod.CASH
    proof_reference: str | None = Field(default=None, max_length=180)
    proof_url: str | None = Field(default=None, max_length=500)
    proof_notes: str | None = Field(default=None, max_length=2000)
    notes: str | None = Field(default=None, max_length=1000)
    idempotency_key: str | None = Field(default=None, min_length=8, max_length=160)

File: products/LoanHub/apps/frontend/api/finance.ts
import type { PaymentMethod } from "@/types/expenseManagement";
import type { PaymentResult } from "@/types/loan";
import type { PaymentProvider } from "@/types/payment";
import { api } from "@/lib/api";

export type FeeType = "flat" | "percentage" | "hybrid";

export type BorrowerRequestFee = {
  id: string;
  company_id: string | null;
  name: string;
  fee_type: FeeType;
  flat_amount: number;
  percentage: number;
  minimum_amount: number | null;
  maximum_amount: number | null;
  currency: string;
  required_before_submission: boolean;
  refundable: boolean;
  effective_from: string | null;
  effective_to: string | null;
  is_active: boolean;
  created_by_user_id: string | null;
  created_at: string;
  updated_at: string;
};

export type BorrowerRequestFeeWrite = Omit<BorrowerRequestFee, "id" | "created_by_user_id" | "created_at" | "updated_at">;

export type AccountOpeningFee = {
  id: string;
  company_id: string | null;
  name: string;
  fee_type: FeeType;
  flat_amount: number;
  percentage: number;
  minimum_amount: number | null;
  maximum_amount: number | null;
  currency: string;
  effective_from: string | null;
  effective_to: string | null;
  is_active: boolean;
  created_by_user_id: string | null;
  created_at: string;
  updated_at: string;
};

export type AccountOpeningFeeWrite = Omit<AccountOpeningFee, "id" | "created_by_user_id" | "created_at" | "updated_at">;

export type TransactionAgreement = {
  id: string;
  company_id: string;
  agreement_number: string;
  name: string;
  inbound_percentage: number;
  outbound_percentage: number;
  inbound_flat_fee: number;
  outbound_flat_fee: number;
  minimum_charge: number | null;
  maximum_charge: number | null;
  currency: string;
  settlement_frequency: "daily" | "weekly" | "monthly" | "quarterly" | "custom";
  settlement_day: number | null;
  effective_from: string;
  effective_to: string | null;
  terms: string | null;
  status: string;
  owner_accepted_by_user_id: string | null;
  company_accepted_by_user_id: string | null;
  owner_accepted_at: string | null;
  company_accepted_at: string | null;
  activated_at: string | null;
  suspended_at: string | null;
  created_at: string;
  updated_at: string;
};

export type TransactionAgreementWrite = Omit<
  TransactionAgreement,
  | "id"
  | "agreement_number"
  | "status"
  | "owner_accepted_by_user_id"
  | "company_accepted_by_user_id"
  | "owner_accepted_at"
  | "company_accepted_at"
  | "activated_at"
  | "suspended_at"
  | "created_at"
  | "updated_at"
>;

export type ChargeLedgerEntry = {
  id: string;
  company_id: string;
  agreement_id: string;
  payment_id: string;
  claim_id: string | null;
  direction: "inbound" | "outbound";
  provider: PaymentProvider;
  payment_purpose: string;
  gross_amount: number;
  percentage_rate: number;
  flat_fee: number;
  charge_amount: number;
  currency: string;
  status: string;
  accrued_at: string;
  claimed_at: string | null;
  settled_at: string | null;
  created_at: string;
};

export type ChargeClaim = {
  id: string;
  company_id: string;
  agreement_id: string;
  claim_number: string;
  period_start: string;
  period_end: string;
  transaction_count: number;
  gross_transaction_value: number;
  amount: number;
  currency: string;
  status: string;
  issued_at: string | null;
  due_at: string | null;
  acknowledged_at: string | null;
  paid_at: string | null;
  disputed_at: string | null;
  notes: string | null;
  dispute_reason: string | null;
  created_at: string;
  updated_at: string;
};

export async function listBorrowerRequestFees(): Promise<BorrowerRequestFee[]> {
  return (await api.get<BorrowerRequestFee[]>("/finance-config/borrower-request-fees")).data;
}

export async function saveBorrowerRequestFee(payload: BorrowerRequestFeeWrite, id?: string): Promise<BorrowerRequestFee> {
  return (await (id
    ? api.put<BorrowerRequestFee>(`/finance-config/borrower-request-fees/${id}`, payload)
    : api.post<BorrowerRequestFee>("/finance-config/borrower-request-fees", payload))).data;
}

export async function listAccountOpeningFees(): Promise<AccountOpeningFee[]> {
  return (await api.get<AccountOpeningFee[]>("/finance-config/account-opening-fees")).data;
}

export async function saveAccountOpeningFee(payload: AccountOpeningFeeWrite, id?: string): Promise<AccountOpeningFee> {
  return (await (id
    ? api.put<AccountOpeningFee>(`/finance-config/account-opening-fees/${id}`, payload)
    : api.post<AccountOpeningFee>("/finance-config/account-opening-fees", payload))).data;
}

export async function listTransactionAgreements(companyId?: string): Promise<TransactionAgreement[]> {
  return (await api.get<TransactionAgreement[]>("/finance-config/agreements", {
    params: companyId ? { company_id: companyId } : undefined,
  })).data;
}

export async function saveTransactionAgreement(payload: TransactionAgreementWrite, id?: string): Promise<TransactionAgreement> {
  return (await (id
    ? api.put<TransactionAgreement>(`/finance-config/agreements/${id}`, payload)
    : api.post<TransactionAgreement>("/finance-config/agreements", payload))).data;
}

export async function makeOwnerAgreementDecision(id: string, accept: boolean, reason?: string): Promise<TransactionAgreement> {
  return (await api.post<TransactionAgreement>(`/finance-config/agreements/${id}/owner-decision`, { accept, reason })).data;
}

export async function makeCompanyAgreementDecision(id: string, accept: boolean, reason?: string): Promise<TransactionAgreement> {
  return (await api.post<TransactionAgreement>(`/finance-config/agreements/${id}/company-decision`, { accept, reason })).data;
}

export async function listChargeLedger(companyId?: string): Promise<ChargeLedgerEntry[]> {
  return (await api.get<ChargeLedgerEntry[]>("/finance-config/charge-ledger", {
    params: companyId ? { company_id: companyId } : undefined,
  })).data;
}

export async function listChargeClaims(companyId?: string): Promise<ChargeClaim[]> {
  return (await api.get<ChargeClaim[]>("/finance-config/claims", {
    params: companyId ? { company_id: companyId } : undefined,
  })).data;
}

export async function createChargeClaim(payload: {
  company_id: string;
  period_start: string;
  period_end: string;
  due_days: number;
  notes?: string | null;
}): Promise<ChargeClaim> {
  return (await api.post<ChargeClaim>("/finance-config/claims", payload)).data;
}

export async function listCompanyAgreements(): Promise<TransactionAgreement[]> {
  return (await api.get<TransactionAgreement[]>("/finance-config/company/agreements")).data;
}

export async function listCompanyChargeLedger(): Promise<ChargeLedgerEntry[]> {
  return (await api.get<ChargeLedgerEntry[]>("/finance-config/company/charge-ledger")).data;
}

export async function listCompanyChargeClaims(): Promise<ChargeClaim[]> {
  return (await api.get<ChargeClaim[]>("/finance-config/company/claims")).data;
}

export async function decideCompanyClaim(id: string, action: "acknowledge" | "dispute", reason?: string): Promise<ChargeClaim> {
  return (await api.post<ChargeClaim>(`/finance-config/claims/${id}/company-decision`, { action, reason })).data;
}

export type ClaimSettlementPayload = {
  payment_method: PaymentMethod;
  proof_reference?: string | null;
  proof_url?: string | null;
  proof_notes?: string | null;
  notes?: string | null;
  idempotency_key?: string;
};

export async function settleCompanyClaim(id: string, payload: ClaimSettlementPayload): Promise<PaymentResult> {
  return (await api.post<PaymentResult>(`/finance-config/claims/${id}/settle`, payload)).data;
}

export const settleCompanyClaimCash = settleCompanyClaim;
